samples/dnn/object_detection.cpp

```diff
@@ -396,13 +396,48 @@ void postprocess(Mat& frame, const std::vector<Mat>& outs, Net& net)
     else
         CV_Error(Error::StsNotImplemented, "Unknown output layer type: " + outLayerType);
 
-    std::vector<int> indices;
-    NMSBoxes(boxes, confidences, confThreshold, nmsThreshold, indices);
-    for (size_t i = 0; i < indices.size(); ++i)
+    if (outLayers.size() > 1)
     {
-        int idx = indices[i];
-        Rect box = boxes[idx];
-        drawPred(classIds[idx], confidences[idx], box.x, box.y,
+        std::map<int, std::vector<size_t> > class2indices;
+        for (size_t i = 0; i < classIds.size(); i++)
+        {
+            if (confidences[i] >= confThreshold)
+            {
+                class2indices[classIds[i]].push_back(i);
+            }
+        }
+        std::vector<Rect> nmsBoxes;
+        std::vector<float> nmsConfidences;
+        std::vector<int> nmsClassIds;
+        for (std::map<int, std::vector<size_t> >::iterator it = class2indices.begin(); it != class2indices.end(); ++it)
+        {
+            std::vector<Rect> local_boxes;
+            std::vector<float> local_confidences;
+            std::vector<size_t> class_indices = it->second;
+            for (size_t i = 0; i < class_indices.size(); i++)
+            {
+                local_boxes.push_back(boxes[class_indices[i]]);
+                local_confidences.push_back(confidences[class_indices[i]]);
+            }
+            std::vector<int> nms_indices;
+            NMSBoxes(local_boxes, local_confidences, confThreshold, nmsThreshold, nms_indices);
+            for (size_t i = 0; i < nms_indices.size(); i++)
+            {
+                size_t idx = nms_indices[i];
+                nmsBoxes.push_back(local_boxes[idx]);
+                nmsConfidences.push_back(local_confidences[idx]);
+                nmsClassIds.push_back(it->first);
+            }
+        }
+        boxes = nmsBoxes;
+        classIds = nmsClassIds;
+        confidences = nmsConfidences;
+    }
+
+    for (size_t i = 0; i < boxes.size(); ++i)
```

Comment:
it'd be perfect to name the iterator as `idx` (so two lines below will be unchanged :) )